iroha/src/sumeragi.rs

```diff
@@ -656,343 +659,561 @@ impl<Q: QueueTrait, G: GenesisNetworkTrait, W: WorldTrait> Sumeragi<Q, G, W> {
+                    n_view_changes: self.n_view_changes,
+                })
+            } else {
+                Err(error!(
+                "Not enough peers to be Byzantine fault tolerant. Expected a least {} peers, got {}",
+                min_peers,
+                peers.len(),
+            ))
+            }
         }
-        let Hash(bytes) = Hash::new(&bytes);
-        let mut rng = StdRng::from_seed(bytes);
-        self.sorted_peers.shuffle(&mut rng);
     }
 
-    /// Shifts `sorted_peers` by one to the right.
-    #[allow(clippy::expect_used)]
-    pub fn shift_peers_by_one(&mut self) {
-        let last_element = self
-            .sorted_peers
-            .pop()
-            .expect("No elements found in sorted peers.");
-        self.sorted_peers.insert(0, last_element);
+    #[derive(Clone, Debug, Encode, Decode)]
+    pub struct Topology {
+        /// Current order of peers. The roles of peers are defined based on this order.
+        sorted_peers: Vec<PeerId>,
+        /// Maximum faulty peers in a network.
+        max_faults: u32,
+
+        reshuffle_after_n_view_changes: u32,
+
+        at_block: Hash,
+
+        n_view_changes: u32,
     }
 
-    /// Shifts `sorted_peers` by `n` to the right.
-    pub fn shift_peers_by_n(&mut self, n: u32) {
-        for _ in 0..n {
-            self.shift_peers_by_one();
+    impl Topology {
+        pub fn builder() -> Builder {
+            Builder::new()
         }
-    }
 
-    /// Get role of the peer by its id.
-    pub fn role(&self, peer_id: &PeerId) -> Role {
-        if self.leader() == peer_id {
-            Role::Leader
-        } else if self.proxy_tail() == peer_id {
-            Role::ProxyTail
-        } else if self.validating_peers().contains(peer_id) {
-            Role::ValidatingPeer
-        } else {
-            Role::ObservingPeer
+        pub fn into_builder(self) -> Builder {
+            Builder {
+                peers: Some(self.sorted_peers.into_iter().collect()),
+                max_faults: Some(self.max_faults),
+                reshuffle_after_n_view_changes: Some(self.reshuffle_after_n_view_changes),
+                at_block: Some(self.at_block),
+                n_view_changes: self.n_view_changes,
+            }
+        }
+
+        pub fn apply_block(&mut self, block_hash: Hash) {
+            *self = self
+                .clone()
+                .into_builder()
+                .at_block(block_hash)
+                .with_view_changes(0)
+                .build()
+                .expect("Given a valid Topology, it is impossible to have error here.")
+        }
+
+        pub fn apply_view_change(&mut self) {
+            let n_view_changes = self.n_view_changes;
+            *self = self
+                .clone()
+                .into_builder()
+                .with_view_changes(n_view_changes + 1)
+                .build()
+                .expect("Given a valid Topology, it is impossible to have error here.")
+        }
```

Comment:
May be we can think of some other way of implementing it, apart from converting into builder and do same data wrangling

Reply:
Ok, its fine i guess as this design will be needed.